### PR TITLE
Fix plugin for 1.8.6

### LINKED ISSUE #3: 2StepAuth breaks login
On MyBB 1.8.6, when logging in with a 2fa-enabled account, this SQL error shows up : 
```
SQL Error:
    1064 - You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 1
Query:
    SELECT * FROM <USERS TABLE> WHERE uid = 
```

### PR DESCRIPTION
**Pull request by s0r00t**
Fixes #3.

### CODE REVIEW

File: plugins/twostepauth.php [hunk 1]
```diff
@@ -971,7 +972,7 @@ function twostepauth_admin_error($title, $msg)
 function twostepauth_authorize_ip($ip, $uid, $code) {
     global $db;
     $loc = twostepauth_get_location($ip);
-    $db->insert_query("twostepauth_authorizations", array("ip" => $db->escape_string($ip), "location" => $db->escape_string($loc), "code" => $db->escape_string($code), "uid" => $db->escape_string($uid)));
+    $db->insert_query("twostepauth_authorizations", array("ip" => $db->escape_string($ip), "location" => $db->escape_string($loc), "code" => $db->escape_string($code), "username" => $db->escape_string($uid)));
```
Comment (Sama34) [inline plugins/twostepauth.php:975]: `$uid` may still actually be a UID (integer).
https://github.com/jariz/2stepauth/pull/4/files#diff-e5b7031cf5da7d3f1ffcd5b5b77c2e69R685

File: plugins/twostepauth.php [hunk 2]
```diff
@@ -697,7 +697,7 @@ function twostepauth_login()
     global $user, $db, $mybb, $footer, $header, $navigation, $headerinclude, $themes, $templates, $usercpnav, $lang, $session, $logins, $twostepauth_loginattempts;
 
     if($user == null) return;
-    $user_info = $db->fetch_array($db->simple_select("users", "*", "uid = ".$user["uid"]));
+    $user_info = $db->fetch_array($db->simple_select("users", "*", "username = '".$user["username"]."'"));
```
Comment (Sama34) [inline plugins/twostepauth.php:700]: Why don't you escape this? Also note that some DB configurations may require you to search on lower case letters.

File: plugins/twostepauth.php [hunk 3]
```diff
@@ -920,12 +920,13 @@ function twostepauth_user_validate($userhandler) {
  * HELPERS
  */
 
-function twostepauth_allowed($uid,$ip=-1) {
+function twostepauth_allowed($username,$ip=-1) {
     global $db;
+    $uid = $db->fetch_field($db->simple_select("users", "uid", "username = '{$username}'"), "uid");
```
Comment (Sama34) [inline plugins/twostepauth.php:925]: And this.

**Comment by s0r00t** [comment]
It seems some bugs are still present : 
-The generated QR code sometimes do not work
-My GAuth codes aren't recognized by the plugin
-And there's the db insert error.

I'll fix those problems. I wrote this patch at 11PM,  so there are some problems.

File: plugins/twostepauth.php (hunk 1 above)
Comment (s0r00t) [inline plugins/twostepauth.php:975]: Not all the time. Sometimes, $user array only contains data from the login form, with no uid. 
However, we could get the uid like I did by fetching the DB.

Comment (Sama34) [inline plugins/twostepauth.php:975]: Yes but in this particular line you are assuming `$uid` is always going to be an `username`, which is not true if you see the line I linked.

**Comment by jariz** [comment]
Hey, any updates? Would like to use it on a forum of mine (lol).

**Comment by s0r00t** [comment]
To be honest, I've spent a lot of time on it, and noticed a looooooooot of things are broken.
It ended up corrupting forums.

It seriously need a code refactor.

**Pull request merged**